src/promptflow/promptflow/executor/_tool_resolver.py

```diff
@@ -293,6 +369,44 @@ def resolve_tool_by_node(self, node: Node, convert_input_types=True) -> Resolved
                 raise ResolveToolError(node_name=node.name, target=e.target, module=e.module) from e
             raise ResolveToolError(node_name=node.name) from e
 
+    def resolve_function_by_definition(self, node_name: str, tool: AssistantTool) -> ResolvedAssistantTool:
+        try:
+            if tool.source is None:
+                raise InvalidAssistantTool(
```

Comment:
I think these validation for non existing fields should be in `from_dict` to simplify the logic here.